src/oop/object_interactions/medium/q4/MainTest.java

```diff
@@ -0,0 +1,197 @@
+                    "You have modified the provided constructor on the " + className + " class. Please revert it to the original state.");
+            char updateType = (char) test[num][2];
+            double updateVolume = (double) test[num][3];
+            Object output = testClass.callMethod(getBloodTypeMethodName, classInstance);
+            _assertEquals(type, output,
+                    "You have modified the provided " + getBloodTypeMethodName + "method  on the " + className + " class. Please revert it to the original state.");
+            output = testClass.callMethod(getVolumeMethodName, classInstance);
+            _assertEquals(volume, output,
+                    "You have modified the provided " + getVolumeMethodName + "method  on the " + className + " class. Please revert it to the original state.");
+            output = testClass.callMethod(setBloodTypeMethodName, new Object[][]{{updateType, char.class}}, classInstance);
+            _assertEquals(updateType, testClass.getFieldValue(classInstance, bloodTypeAttributeName),
+                    "You have modified the provided " + setBloodTypeMethodName + "method  on the " + className + " class. Please revert it to the original state.");
+            assertNull(output, String.join(" ", "Your", setBloodTypeMethodName, "should not return any output"));
+            output = testClass.callMethod(setVolumeMethodName, new Object[][]{{updateVolume, double.class}}, classInstance);
+            _assertEquals(updateVolume, testClass.getFieldValue(classInstance, volumeAttributeName),
+                    "You have modified the provided " + setVolumeMethodName + "method  on the " + className + " class. Please revert it to the original state.");
+            assertNull(output, String.join(" ", "Your", setVolumeMethodName, "should not return any output"));
+        }
+    }
+
+    private static Stream<Arguments> inputProvider() {
+        return Stream.of(
+                Arguments.of('A', 10.3, 'B', 4.9),
+                Arguments.of('A', 10.3, 'A', 7.2),
+                Arguments.of('A', 10.3, 'A', minBlood),
+                Arguments.of('A', 10.3, 'A', minBlood - 1),
+                Arguments.of('A', 10.3, 'A', minBlood + 1),
+                Arguments.of('B', 4.2, 'A', 2.5),
+                Arguments.of('O', 3.4, 'B', 5.7)
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("inputProvider")
+    @Tag("dependent2")
+    public void correctTransfusionFromMethod(char type, double volume, char donType, double donVolume) throws Throwable {
+        Object[][] arguments = {
+                {type, char.class},
+                {volume, double.class}
+        };
+        Snail donor = new Snail(donType, donVolume);
+        Object classInstance = testClass.createInstance(arguments);
+        Object[][] methodArguments = {
+                {donor, Snail.class}
+        };
+        Object output;
```

Comment:
use a .hasMethod before the if statements with the .callMethod. This will also get rid of the need for an output variable as assertNull will not need to be checked